src/borrow.rs

```diff
@@ -0,0 +1,168 @@
+use std::cell::UnsafeCell;
+use std::collections::hash_map::{Entry, HashMap};
```

Comment:
The performance overhead could possibly be reduced by using an AHash-backed hash table here instead of the SipHash-based one shipped by the standard library if the additional dependency is considered worth it.

Reply:
Let's not worry about this just yet...